### ISSUE TITLE
[Need support]: New element locator have not been found. There is a lack of reference data.

### ISSUE THREAD

**Issue by hemanth-tr** [healenium/healenium-web#185]
### Describe the problem

I ran the pure tests first and made changes to element id and ran the tests again, unfortunately tests are failed. When i took a look at the proxy logs i noticed following log.

WARN 3940 --- [nio-8085-exec-4] c.e.h.processor.HealingProcessor         : New element locator have not been found. There is a lack of reference data.

### Healenium Web version

None

### Healenium Backend version

1.0

### Selenium version

4.1.1

### Platform

.NET

### Properties file

There is no properties file when we run tests on **.NET**. Note: Needed components are installed in a server and RemoteWebDriver is pointed to that server **healenium-proxy**.

http://{server-ip}:8085

Note: _application.yml_ in **healenium-proxy** project is updated accordingly

### Logs appeared during using Healenium

```shell
2022-06-14 05:45:41.357  WARN 3940 --- [nio-8085-exec-4] c.e.h.processor.FindElementProcessor     : Failed to find an element using locator By.cssSelector: #UserName
Reason: no such element: Unable to locate element: {"method":"css selector","selector":"#UserName"}
  (Session info: chrome=102.0.5005.115)
For documentation on this error, please visit: https://selenium.dev/exceptions/#no_such_element
Build info: version: '4.1.0', revision: '87802e897b'
System info: host: '{server-name}', ip: '{server-ip}', os.name: 'Windows Server 2012 R2', os.arch: 'amd64', os.version: '6.3', java.version: '18.0.1.1'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [40bd3e6049ff6e4158dba5974c12b75f, findElement {using=css selector, value=#UserName}]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 102.0.5005.115, chrome: {chromedriverVersion: 102.0.5005.61 (0e59bcc00cc4..., userDataDir: C:\Users\SUSHRI~1\AppData\L...}, goog:chromeOptions: {debuggerAddress: localhost:55602}, javascriptEnabled: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: WINDOWS, platformName: WINDOWS, proxy: {}, se:cdp: ws://{server-ip}:4444/sess..., se:cdpVersion: 102.0.5005.115, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
Session ID: 40bd3e6049ff6e4158dba5974c12b75f
Trying to heal...
2022-06-14 05:45:41.877  WARN 3940 --- [nio-8085-exec-4] c.e.h.processor.HealingProcessor         : New element locator have not been found. There is a lack of reference data.
2022-06-14 05:45:41.879 ERROR 3940 --- [nio-8085-exec-4] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.openqa.selenium.NoSuchElementException: no such element: Unable to locate element: {"method":"css selector","selector":"#UserName"}
  (Session info: chrome=102.0.5005.115)
For documentation on this error, please visit: https://selenium.dev/exceptions/#no_such_element
Build info: version: '4.1.0', revision: '87802e897b'
System info: host: '{server-name}', ip: '{server-ip}', os.name: 'Windows Server 2012 R2', os.arch: 'amd64', os.version: '6.3', java.version: '18.0.1.1'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [40bd3e6049ff6e4158dba5974c12b75f, findElement {using=css selector, value=#UserName}]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 102.0.5005.115, chrome: {chromedriverVersion: 102.0.5005.61 (0e59bcc00cc4..., userDataDir: C:\Users\SUSHRI~1\AppData\L...}, goog:chromeOptions: {debuggerAddress: localhost:55602}, javascriptEnabled: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: WINDOWS, platformName: WINDOWS, proxy: {}, se:cdp: ws://{server-ip}:4444/sess..., se:cdpVersion: 102.0.5005.115, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
Session ID: 40bd3e6049ff6e4158dba5974c12b75f] with root cause

org.openqa.selenium.NoSuchElementException: no such element: Unable to locate element: {"method":"css selector","selector":"#UserName"}
  (Session info: chrome=102.0.5005.115)
For documentation on this error, please visit: https://selenium.dev/exceptions/#no_such_element
Build info: version: '4.1.0', revision: '87802e897b'
System info: host: '{server-name}', ip: '{server-ip}', os.name: 'Windows Server 2012 R2', os.arch: 'amd64', os.version: '6.3', java.version: '18.0.1.1'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [40bd3e6049ff6e4158dba5974c12b75f, findElement {using=css selector, value=#UserName}]
Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 102.0.5005.115, chrome: {chromedriverVersion: 102.0.5005.61 (0e59bcc00cc4..., userDataDir: C:\Users\SUSHRI~1\AppData\L...}, goog:chromeOptions: {debuggerAddress: localhost:55602}, javascriptEnabled: true, networkConnectionEnabled: false, pageLoadStrategy: normal, platform: WINDOWS, platformName: WINDOWS, proxy: {}, se:cdp: ws://{server-ip}:4444/sess..., se:cdpVersion: 102.0.5005.115, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
Session ID: 40bd3e6049ff6e4158dba5974c12b75f
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:483) ~[na:na]
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.createException(W3CHttpResponseCodec.java:200) ~[selenium-remote-driver-4.1.0.jar:na]
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:133) ~[selenium-remote-driver-4.1.0.jar:na]
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:53) ~[selenium-remote-driver-4.1.0.jar:na]
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:184) ~[selenium-remote-driver-4.1.0.jar:na]
	at com.epam.healenium.healenium_proxy.command.HealeniumCommandExecutor.execute(HealeniumCommandExecutor.java:28) ~[classes/:na]
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:559) ~[selenium-remote-driver-4.1.0.jar:na]
	at org.openqa.selenium.remote.ElementLocation$ElementFinder$2.findElement(ElementLocation.java:162) ~[selenium-remote-driver-4.1.0.jar:na]
	at org.openqa.selenium.remote.ElementLocation.findElement(ElementLocation.java:60) ~[selenium-remote-driver-4.1.0.jar:na]
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:383) ~[selenium-remote-driver-4.1.0.jar:na]
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:375) ~[selenium-remote-driver-4.1.0.jar:na]
	at com.epam.healenium.processor.FindElementProcessor.execute(FindElementProcessor.java:24) ~[healenium-web-3.2.4.jar:na]
	at com.epam.healenium.processor.BaseProcessor.process(BaseProcessor.java:41) ~[healenium-web-3.2.4.jar:na]
	at com.epam.healenium.handlers.proxy.BaseHandler.findElement(BaseHandler.java:58) ~[healenium-web-3.2.4.jar:na]
	at com.epam.healenium.handlers.proxy.SelfHealingProxyInvocationHandler.invoke(SelfHealingProxyInvocationHandler.java:39) ~[healenium-web-3.2.4.jar:na]
	at jdk.proxy2/jdk.proxy2.$Proxy62.findElement(Unknown Source) ~[na:na]
	at com.epam.healenium.healenium_proxy.request.post.override.HealeniumFindElementPostRequest.getHealingResponse(HealeniumFindElementPostRequest.java:116) ~[classes/:na]
	at com.epam.healenium.healenium_proxy.request.post.override.HealeniumFindElementPostRequest.execute(HealeniumFindElementPostRequest.java:89) ~[classes/:na]
	at com.epam.healenium.healenium_proxy.request.post.HealeniumPostRequest.execute(HealeniumPostRequest.java:41) ~[classes/:na]
	at com.epam.healenium.healenium_proxy.handler.HealeniumProxyHttpHandler.handleRequest(HealeniumProxyHttpHandler.java:23) ~[classes/:na]
	at org.springframework.web.servlet.mvc.HttpRequestHandlerAdapter.handle(HttpRequestHandlerAdapter.java:52) ~[spring-webmvc-5.3.15.jar:5.3.15]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067) ~[spring-webmvc-5.3.15.jar:5.3.15]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963) ~[spring-webmvc-5.3.15.jar:5.3.15]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[spring-webmvc-5.3.15.jar:5.3.15]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909) ~[spring-webmvc-5.3.15.jar:5.3.15]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:681) ~[tomcat-embed-core-9.0.56.jar:4.0.FR]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[spring-webmvc-5.3.15.jar:5.3.15]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764) ~[tomcat-embed-core-9.0.56.jar:4.0.FR]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.3.15.jar:5.3.15]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.15.jar:5.3.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.3.15.jar:5.3.15]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.15.jar:5.3.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.3.15.jar:5.3.15]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.15.jar:5.3.15]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:540) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:382) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:895) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1732) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-9.0.56.jar:9.0.56]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]

2022-06-14 05:45:42.691  INFO 3940 --- [nio-8085-exec-5] c.e.h.h.r.delete.HealeniumDeleteRequest  : Report available at http://localhost:7878/healenium/report/40bd3e6049ff6e4158dba5974c12b75f
```

**Comment by ElenaStepuro** [comment]
Hello @hemanth-tr , please, update backend version to 3.2.2 and proxy  version to 0.2.5. Version 1.0 doesn't support Selenium 4

**Comment by ElenaStepuro** [comment]
Here you can find repository for backend with latest changes: https://github.com/healenium/healenium-backend
And here for proxy: https://github.com/healenium/healenium-proxy

**Comment by hemanth-tr** [comment]
@ElenaStepuro i did that in docker first before doing it on server, seems like elements are not healed yet. :(

![image](https://user-images.githubusercontent.com/24373661/173591281-73121ae6-7eb3-49a5-8d38-c28d1a8fd933.png)

**Comment by Aliaksei-Ashukha** [comment]
Hello @hemanth-tr,
"New element locator have not been found. There is a lack of reference data." means etalon data has not survived after pure tests. 
Kindly check data appropriate locator in the "selector" table of postgres database.

**Comment by hemanth-tr** [comment]
> Hello @hemanth-tr, "New element locator have not been found. There is a lack of reference data." means etalon data has not survived after pure tests. Kindly check data appropriate locator in the "selector" table of postgres database.

Hi @Aliaksei-Ashukha,

I took a look at posgresql database and found that each time we run the test it is a whole new set of data for healenium because, our landing page url contains varying data i.e, timestamp. So, when the element identification is changed and we run the test, selenium would not find the element obviously and healenium wouldn't either because there is no selector exists at this point as timestamp in url is changed. Is there anyway to get around this?

**Comment by Aliaksei-Ashukha** [comment]
Hi @hemanth-tr ,
Thanks for your details. We understood your problem. We are assessing the one. We let you know about results and time frame.

**Comment by hemanth-tr** [comment]
Thanks @Aliaksei-Ashukha , appreciate your quick response on this.

**Comment by mayukh-kant** [comment]
> Hi @hemanth-tr , Thanks for your details. We understood your problem. We are assessing the one. We let you know about results and time frame.

+1, We have the same problem.

**Comment by Aliaksei-Ashukha** [comment]
Hi @hemanth-tr , @mayukh-kant 
Issue was fixed at hlm-backend 3.2.4: https://github.com/healenium/healenium-web/issues/204#issuecomment-1231792830
